spring-data-microservice-transaction-sample/README.md

```diff
@@ -0,0 +1,548 @@
+
+At first, Order Service starts a two-phase commit transaction by calling the repository class's [executeTwoPcTransaction()](order-service/src/main/java/sample/order/OrderService.java#L100-L102):
+
+```java
+execAndReturnResponse(responseObserver, "Placing an order", () -> {
+  // Start a two-phase commit transaction
+  TwoPcResult<String> result = orderRepository.executeTwoPcTransaction(txId -> {
+```
+
+The following `Execute CRUD operations`, `Two-phase Commit` and `Error handling` are automatically performed inside the API.
+
+### 2. Execute CRUD operations
+
+After the transaction is started, the CRUD operations are executed by `executeTwoPcTransaction()`.
+
+Order Service puts the order information to the `order_service.orders` table also the detailed information to `order_service.statements` (the code is [here](order-service/src/main/java/sample/order/OrderService.java#L106-L128)):
+
+```java
+// Put the order info into the orders table
+orderRepository.insert(order);
+
+AtomicInteger amount = new AtomicInteger();
+for (ItemOrder itemOrder : request.getItemOrderList()) {
+  int itemId = itemOrder.getItemId();
+  int count = itemOrder.getCount();
+  // Retrieve the item info from the items table
+  Optional<Item> itemOpt = itemRepository.findById(itemId);
+  if (!itemOpt.isPresent()) {
+    String message = "Item not found: " + itemId;
+    responseObserver.onError(
+        Status.NOT_FOUND.withDescription(message).asRuntimeException());
+    throw new ScalarDbNonTransientException(message);
+  }
+  Item item = itemOpt.get();
+
+  int cost = item.price * count;
+  // Put the order statement into the statements table
+  statementRepository.insert(new Statement(itemId, orderId, count));
+  // Calculate the total amount
+  amount.addAndGet(cost);
+}
+```
+
+And, Order Service calls the `payment` gRPC endpoint of Customer Service along with the transaction ID (the code is [here](order-service/src/main/java/sample/order/OrderService.java#L131)).
+
+This endpoint first joins the transaction with [join()](customer-service/src/main/java/sample/customer/CustomerService.java#L183):
+
+```java
+if (isJoin) {
+  // Join the transaction
+  customerRepository.join(txId);
+} else {
+  // Resume the transaction
+  customerRepository.resume(txId);
+}
+```
+
+`join()` is called via [execTwoPcOperation()](customer-service/src/main/java/sample/customer/CustomerService.java#L99):
+
+```java
+public void payment(PaymentRequest request, StreamObserver<Empty> responseObserver) {
+  execTwoPcOperation(request.getTransactionId(), true, responseObserver, "Payment", () -> {
+```
+
+It then gets the customer information, and checks if the customer's credit total exceeds the credit limit after the payment.
+And if the check is okay, it updates the customer's credit total (the code is [here](customer-service/src/main/java/sample/customer/CustomerService.java#L100-L114)):
```

Comment:
In these two sentences, I assumed that "it" meant "endpoint. Could you please confirm if that is correct?

```suggestion
The endpoint then gets the customer information, and checks if the customer's credit total exceeds the credit limit after the payment.
If the check is okay, the endpoint updates the customer's credit total (the code is [here](customer-service/src/main/java/sample/customer/CustomerService.java#L100-L114)):
```